### ISSUE TITLE
Redundant file/web hits for cached documents

### ISSUE THREAD

**Issue by faramos** [microsoft/ClearScript#219]
Using V8ScriptEngine in a context where there are several process definitions in different versions, where each one has a module and several expressions that use this module.
I see that we are in need of a performance tuning.

I was wondering if adding a system document makes it compile, because I don't see how I can add an already compiled one.
Currently, I´m adding it in a scoped service this way:
```
engine.DocumentSettings.AddSystemDocument("common", ModuleCategory.Standard, jsCommon);
```

As for expressions, I plan to cache these compiled.

Thanks in advance.

**Comment by ClearScriptLib** [comment]
Hi @faramos,

>I was wondering if adding a system document makes it compile, because I don't see how I can add an already compiled one.

All script code must be compiled before V8 can execute it. The `V8ScriptEngine` methods that execute script code directly simply perform compilation as an internal step.

That's probably not very helpful, so let's consider your example. The first time `common` is imported, it's compiled, executed, and cached in its compiled form. Subsequent imports reuse the cached module.

Note that modules can be shifted out of the cache. The limit is adjustable via [`ModuleCategory.Standard.MaxCacheSize`](https://microsoft.github.io/ClearScript/Reference/html/P_Microsoft_ClearScript_DocumentCategory_MaxCacheSize.htm) and has a default value of 1024. Unfortunately `V8ScriptEngine` instances can't share modules even if they share a V8 runtime.

Note also that regular scripts and CommonJS modules are also cached and reused whenever possible, so implementing your own cache is probably unnecessary. And unlike standard modules, these document categories _are_ shared across engine instances within the same runtime.

Good luck!

**Comment by ClearScriptLib** [comment]
Please reopen this issue if you have additional questions or comments. Thanks!

**Comment by faramos** [comment]
I have come back on the subject and I was debugging the default loader.
My concern is that even though the document is cached, it still makes a request on every import to see if the document exists.

At Microsoft.ClearScript.DocumentLoader.DefaultImpl

`public override async Task<Document> LoadDocumentAsync(DocumentSettings settings, DocumentInfo? sourceInfo, string specifier, DocumentCategory category, DocumentContextCallback contextCallback)`

first analyzes the candidate uris, sending a http get in case of uris, and then load the document from de cache.

Below I copy a possible workaround to avoid this request.

![image](https://user-images.githubusercontent.com/1985827/126526383-7e5608b7-4ca7-437f-b859-78d851b032ea.png)

**Comment by ClearScriptLib** [comment]
Hi @faramos,

Confirmed! Thank you very much for making us aware of this issue. We'll fix it in the next release, which should be out shortly.

Cheers!

**Comment by ClearScriptLib** [comment]
Fixed in [Version 7.1.5](https://github.com/microsoft/ClearScript/releases/tag/7.1.5).
